lib/galaxy/tools/cache.py

```diff
@@ -288,20 +287,20 @@ class ToolShedRepositoryCache:
     repos_by_tuple: Dict[Tuple[str, str, str], List[ToolConfRepository]]
 
     def __init__(self, session: sessionmaker):
-        engine = session().bind
-        self.session = scoped_session(sessionmaker(engine))
+        self.session = session()
         # Contains ToolConfRepository objects created from shed_tool_conf.xml entries
         self.local_repositories = []
         # Repositories loaded from database
         self.repositories = []
         self.repos_by_tuple = defaultdict(list)
-        self.rebuild()
+        self._rebuild()
+        self.session.close()
```

Comment:
Given the `defer` clause in `_rebuild`. `session.close()` expunges the session, so you should get a DetachedInstanceError when trying to access the `ToolShedRepository.metadata` property. If we call `_rebuild` before accessing the cache, then it will repopulate the session, so that's fine. So this would work only if:
a) we do not access the `metadata` property, or
b) we call `_rebuild` before accessing it.

I'm not sure about the context in which we access the cache, but would it make sense to *not* call `self._rebuild()` from the constructor, and unconditionally rebuild on first access? (that could be done with a property that would lazy-load `self.repositories` maybe?).

Reply:
It'll never be accessed, the cache is disposed before the app startup is finished.